Fix #4315: Optimize dynamicImports of native imports

gzm0 opened the pull request.

gzm0 commented:
CI for now.

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/OptimizerCore.scala:
```diff
@@ -5434,16 +5590,58 @@ private[optimizer] object OptimizerCore {
+      }
+
+      val jsDynImportThunkFor = body match {
+        case Apply(_, New(clazz, _, _), MethodIdent(target), _) if clazz == enclosingClassName =>
+          Some(target)
```
IIUC, when this case applies, we will ignore the constructor of `clazz` and its superclasses. Either we need to check that they are side-effect-free somehow, or (probably easier), we should preserve the `New`, and then `transformStat` it somewhere down the line (which, when the class is inlineable and its constructor is indeed side-effect-free, will eventually collapse to `Skip()`).

gzm0 replied:
I'm not sure I understand. This is just a heuristics of whether we should even *try* to inline the `ApplyDynamicImport`. The full inlining machinery is used to actually inline (or not).

gzm0 commented on linker/shared/src/test/scala/org/scalajs/linker/OptimizerTest.scala:
```diff
@@ -288,12 +288,67 @@ class OptimizerTest {
+        }
+      }
+
+      main.methods.foreach(v => traverser.traverseMemberDef(v.value))
```
Can't... `traverseClassDef` traverses a `ClassDef`. But we have a `LinkedClass` here (or am I missing something?)

sjrd replied:
Oh right. I got bitten by that as well not long ago.

gzm0 commented:
I have
- addressed all comments except the ones regarding pretransforms / rollback machinery.
- rebased past #4647

Note that currently, this PR doesn't compile, since the rebase on #4647 is non trivial.

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/OptimizerCore.scala:
```diff
@@ -5444,17 +5434,15 @@ private[optimizer] object OptimizerCore {
-        isForwarder: Boolean
-    )
-  }
+  final class MethodAttributes private[OptimizerCore] (
```
Made it a case class.

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/OptimizerCore.scala:
```diff
@@ -4780,6 +4895,33 @@ private[optimizer] object OptimizerCore {
+  private final class ImportReplacement(target: ImportTarget,
+      moduleVarName: LocalName, path: List[String]) {
+
+    private[this] var used: Boolean = false
+    private[this] var failed: Boolean = false
```
Yes. At first I thought that the entire rollback machinery is only used on a single trampoline. But that is not the case: We specifically have infrastructure to roll-back across trampolines. So indeed: This is plain wrong and needs to be fixed :)

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/OptimizerCore.scala:
```diff
@@ -1726,6 +1770,102 @@ private[optimizer] abstract class OptimizerCore(config: CommonPhaseConfig) {
+              def genThen(receiver: Tree, expr: Tree) =
+                JSMethodApply(receiver, StringLiteral("then"), List(expr))
+
+              val importTree = config.coreSpec.moduleKind match {
+                case ModuleKind.NoModule =>
+                  throw new AssertionError("ApplyDynamicImport in NoModule mode")
+
+                case ModuleKind.ESModule =>
+                  JSImportCall(StringLiteral(jsNativeLoadSpec.module))
+
+                case ModuleKind.CommonJSModule =>
+                  val require =
+                    JSFunctionApply(JSGlobalRef("require"), List(StringLiteral(jsNativeLoadSpec.module)))
+
+                  val unitPromise = JSMethodApply(
+                    JSGlobalRef("Promise"), StringLiteral("resolve"), List(Undefined()))
+                  genThen(unitPromise, Closure(arrow = true, Nil, Nil, None, require, Nil))
+              }
```
I'd like to factor this out, but couldn't really wrap my head around how tree generation helpers are called in OptimizerCore. Seems like objects with apply methods are most common (e.g. `AndThen`).

Should I define an `object JSImport` with an apply method that does this? (and `object Then`?)

sjrd replied:
That sounds reasonable. I guess we went the `apply` route because we sometimes need the `unapply` as well. Perhaps we even had `AndThen.unapply` in the past, and it became irrelevant later?

In any case, defining a new object near `AndThen` shouldn't hurt.

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/OptimizerCore.scala:
```diff
@@ -4845,27 +5007,40 @@ private[optimizer] object OptimizerCore {
+  private class Scope private (
+    val env: OptEnv,
+    val implsBeingInlined: Set[Scope.InliningID],
+    val importReplacement: Option[ImportReplacement]
```
I'm not sure the Scope (as opposed to the Env) is the right place to put this.

My understanding at some point was:
- `OptEnv`: IR environment *before* optimizer transformations.
- `Scope`: JavaScript scope that is being constructed.

Notably, it seems the scope travels across Closure boundaries, wheres the `OptEnv` does not.

Is this understanding correct? It somewhat feels I should be able to re-use the `Binding` infrastructure for this (also: now that I think more about this: JS scopes are not really meaningful at this level, since all references we produce must be valid in IR scope again).

sjrd replied:
> Notably, it seems the scope travels across Closure boundaries, wheres the `OptEnv` does not.

Yes, that is correct. There's no real distinction beyond that. Currently it's used to track the set of methods being inlined at the current point, so that we never inline a method within itself, even inside a `Closure` (because the method body could contain a `Closure` whose body calls the method; we must not inline in those situations).

If the information you're adding should cross `Closure` boundaries, it should go in `Scope`. Otherwise, it should go in `OptEnv`.

gzm0 commented:
I believe I have addressed all the comments. I'm not sure if the trailing commits are meaningful. LMK if you'd like me to squash them.

sjrd reviewed (changes requested):
Only minor details.

Sorry for the delay here. I was mostly taken by our other issue.

sjrd commented:
> I'm not sure if the trailing commits are meaningful. LMK if you'd like me to squash them.

I think all the commits but the first one can be squashed together, yes.
